crates/noirc_evaluator/src/ssa_refactor/opt/unrolling.rs

```diff
@@ -0,0 +1,575 @@
+    match function.dfg[block].terminator() {
+        Some(TerminatorInstruction::Jmp { arguments, .. }) => {
+            assert_eq!(arguments.len(), 1);
+            let value = arguments[0];
+            if function.dfg.get_numeric_constant(value).is_some() {
+                Ok(value)
+            } else {
+                Err(())
+            }
+        }
+        _ => Err(()),
+    }
+}
+
+/// Unrolls the header block of the loop. This is the block that dominates all other blocks in the
+/// loop and contains the jmpif instruction that lets us know if we should continue looping.
+/// Returns Some(iteration context) if we should perform another iteration.
+fn unroll_loop_header<'a>(
+    function: &'a mut Function,
+    loop_: &'a Loop,
+    unroll_into: BasicBlockId,
+    induction_value: ValueId,
+) -> Option<LoopIteration<'a>> {
+    // We insert into a fresh block first and move instructions into the unroll_into block later
+    // only once we verify the jmpif instruction has a constant condition. If it does not, we can
+    // just discard this fresh block and leave the loop unmodified.
+    let fresh_block = function.dfg.make_block();
+
+    let mut context = LoopIteration::new(function, loop_, fresh_block, loop_.header);
+    let source_block = &context.function.dfg[context.source_block];
+    assert_eq!(source_block.parameters().len(), 1, "Expected only 1 argument in loop header",);
+
+    // Insert the current value of the loop induction variable into our context.
+    let first_param = source_block.parameters()[0];
+    context.values.insert(first_param, induction_value);
+    context.inline_instructions_from_block();
+
+    match context.function.dfg[fresh_block].unwrap_terminator() {
+        TerminatorInstruction::JmpIf { condition, then_destination, else_destination } => {
+            let next_blocks = context.handle_jmpif(*condition, *then_destination, *else_destination);
+
+            // If there is only 1 next block the jmpif evaluated to a single known block.
+            // This is the expected case and lets us know if we should loop again or not.
+            if next_blocks.len() == 1 {
+                let mut instructions = std::mem::take(context.function.dfg[fresh_block].instructions_mut());
+                let terminator = context.function.dfg[fresh_block].take_terminator();
+                context.function.dfg[unroll_into].instructions_mut().append(&mut instructions);
+                context.function.dfg.set_block_terminator(unroll_into, terminator);
+
+                loop_.blocks.contains(&context.source_block).then_some(context)
+            } else {
+                // If this case is reached the loop either uses non-constant indices or we need
+                // another pass, such as mem2reg to resolve them to constants.
+                None
+            }
+        }
+        other => panic!("Expected loop header to terminate in a JmpIf to the loop body, but found {other:?} instead"),
```

Comment:
small nit: I tend to prefer using `unreachable` or `todo` since they indicate intent a bit better -- not a blocker though